cpp/src/arrow/compute/kernels/vector_replace.cc

```
@@ -442,23 +442,406 @@ struct ReplaceWithMaskFunctor {
+    output->length = array.length;
+    int8_t direction = 1;
+
+    if (array.MayHaveNulls()) {
+      ARROW_ASSIGN_OR_RAISE(
+          auto null_bitmap,
+          arrow::internal::CopyBitmap(ctx->memory_pool(), array.buffers[0]->data(),
+                                      array.offset, array.length));
+      return FillNullExecutor<Type>::ExecFillNull(ctx, array, null_bitmap->data(), output,
+                                                  direction, last_valid_value_chunk,
+                                                  last_valid_value_offset);
+    } else {
+      if (array.length > 0) {
+        *last_valid_value_offset = array.length - 1;
+      }
+      *output = array;
+    }
+    return Status::OK();
+  }
+
+  static Status FillNullForwardChunkedArray(KernelContext* ctx,
+                                            const std::shared_ptr<ChunkedArray>& values,
+                                            Datum* out) {
+    if (values->null_count() == 0) {
+      *out = Datum(values);
+      return Status::OK();
+    }
+    if (values->null_count() == values->length()) {
+      *out = Datum(values);
+      return Status::OK();
+    }
+
+    ArrayVector new_chunks;
+    if (values->length() > 0) {
+      ArrayData* array_with_current = &*values->chunk(/*first_chunk=*/0)->data();
+      int64_t last_valid_value_offset = -1;
+
+      for (const auto& chunk : values->chunks()) {
+        if (is_fixed_width(out->type()->id())) {
+          auto* output = out->mutable_array();
+          auto bit_width = checked_cast<const FixedWidthType&>(*output->type).bit_width();
+          auto data_bytes = bit_util::BytesForBits(bit_width * chunk->length());
+          ARROW_ASSIGN_OR_RAISE(output->buffers[0], ctx->AllocateBitmap(chunk->length()));
+          ARROW_ASSIGN_OR_RAISE(output->buffers[1], ctx->Allocate(data_bytes));
+        }
+        DCHECK_OK(FillNullForwardArray(ctx, *chunk->data(), out, *array_with_current,
```

Do not use DCHECK_OK for this. Use RETURN_NOT_OK.

---

DCHECK_OK will crash the program if an error occurs, instead of returning it to the caller.

---

We use it below for kernel registration because we expect that to never fail, but we do expect kernels themselves to fail normally.